docs(plugins): more in-depth pluginOptionsSchema documentation (#27541)
gatsbyjs/gatsby: docs/docs/configuring-usage-with-plugin-options.md

    optionB: Joi.boolean().description(`Enables optionB.`),
  })
}
```

Accessing `pluginOptions.message` would then log `"default message"` in all plugin APIs if the user does not supply their own value.

#### Validate external access

Some plugins (particularly source plugins) query external APIs. With the [`.external()`](https://joi.dev/api/?v=17.3.0#anyexternalmethod-description) method, you can asynchronously validate that the user has access to the API, providing a better experience if they pass invalid secrets.

For example, this is how the [Contentful source plugin](/plugins/gatsby-source-contentful/) might validate that the user has access to the space they are trying to query:

```javascript:title=gatsby-source-contentful/gatsby-node.js
exports.pluginOptionsSchema = ({ Joi }) => {
  return Joi.object({
    accessToken: Joi.string().required(),
    spaceId: Joi.string().required(),
    // ...more options here...
  }).external(async pluginOptions => {
    try {
      await contentful
        .createClient({
          space: pluginOptions.spaceId,
          accessToken: pluginOptions.accessToken,
        })
        .getSpace()

>     } catch (err) {

Reviewer: ```suggestion
    } catch {
```

      throw new Error(
        `Cannot access Contentful space "${pluginOptions.spaceId}" with the provided access token. Double check they are correct and try again!`
      )
    }
  })
}
```

#### Add custom error messages

> Sometimes you might want to provide more detailed error messages when validation fails for a specific field. Joi provides a [`.messages()` method](https://joi.dev/api/?v=17.2.1#anymessagesmessages) which lets you override error messages for specific [error types](https://joi.dev/api/?v=17.2.1#list-of-errors) (e.g. `"any.required"` when a `.required()` call fails).

Reviewer: ```suggestion
Sometimes you might want to provide more detailed error messages when validation fails for a specific field. Joi provides a [`.messages()` method](https://joi.dev/api/?v=17.3.0#anymessagesmessages) which lets you override error messages for specific [error types](https://joi.dev/api/?v=17.3.0#list-of-errors) (e.g. `"any.required"` when a `.required()` call fails).
```

For example, in the `gatsby-plugin-console-log` plugin above, this is how you would provide a custom error message if users do not specify `optionA`:

```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
exports.pluginOptionsSchema = ({ Joi }) => {
  return Joi.object({
    optionA: Joi.boolean()
      .required()
      .description(`Enables optionA.`)
      // highlight-start
      .messages({
        // Override the error message if the .required() call fails
        "any.required": `"optionA" needs to be specified to true or false. Get the correct value from your dashboard settings.`,
      }),
    // highlight-end
    message: Joi.string()
      .default(`default message`)
      .description(`The message logged to the console.`),
    optionB: Joi.boolean().description(`Enables optionB.`),
  })
}
```

#### Deprecating options

> While you can simply remove options from the schema in major versions, that causes cryptic error messages for users upgrading with existing configuration. Instead, deprecate them using the [`.forbidden()` method](https://joi.dev/api/?v=17.2.1#anyforbidden) in a major version release. Then, [add a custom error message](#add-custom-error-messages) explaining how users should upgrade the functionality using `.messages()`.

Reviewer: ```suggestion
While you can simply remove options from the schema in major versions, that causes cryptic error messages for users upgrading with existing configuration. Instead, deprecate them using the [`.forbidden()` method](https://joi.dev/api/?v=17.3.0#anyforbidden) in a major version release. Then, [add a custom error message](#add-custom-error-messages) explaining how users should upgrade the functionality using `.messages()`.
```

For example:

```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
  return Joi.object({
    optionA: Joi.boolean()
      .required()
      .description(`Enables optionA.`)
      // highlight-start
      .forbidden()
      .messages({
        // Override the error message if the .forbidden() call fails
        "any.unknown": `"optionA" is no longer supported. Use "optionB" instead by setting it to the same value you had before on "optionA".`,
      }),
      // highlight-end
    message: Joi.string()
      .default(`default message`)
      .description(`The message logged to the console.`),
    optionB: Joi.boolean().description(`Enables optionB.`),
  })
}
```

### Unit testing an options schema

To verify that a `pluginOptionsSchema` behaves as expected, unit test it with different configurations using the [`gatsby-plugin-utils` package](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-utils#testpluginoptionsschema).

1. Add the `gatsby-plugin-utils` package to your site:

   ```shell

>    npm install gatsby-plugin-utils

Reviewer: This should be a `devDependency` I think? 🤔

```suggestion
   npm install --dev gatsby-plugin-utils
```

You want users to pass in a boolean to `optionA` and a string to `message`, and they can optionally pass a boolean to `optionB`. To enforce these rules, you would create the following `pluginOptionsSchema`:

```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
exports.pluginOptionsSchema = ({ Joi }) => {
  return Joi.object({
    optionA: Joi.boolean().required().description(`Enables optionA.`),
    message: Joi.string()
      .required()
      .description(`The message logged to the console.`),
    optionB: Joi.boolean().description(`Enables optionB.`),
  })
}
```

If users pass options that do not match the schema, the validation will show an error when they run `gatsby develop` and prompt them to fix their configuration.

For example, if an integer is passed into `message` (which is marked as a required string) this message would be shown:

```
 ERROR #11331  PLUGIN

Invalid plugin options for "gatsby-plugin-console-log":

- "message" must be a string
```
### Best practices for option schemas

The [Joi API documentation](https://joi.dev/api/) is a great reference to use while working on a `pluginOptionsSchema`, as it shows all the available types and methods.

> Here are some specific Joi best practices for `pluginOptionsSchema`s:

Reviewer 2: Having the trailing "s" looked funny to me

```suggestion
Here are some specific Joi best practices for `pluginOptionsSchema`:
```

> 

Reviewer 2: Upon rereading this doc, it might make more sense for the "what can be passed in as options" section to come before the "where to access plugin options" section. So that it flows from why you may want plugin options → what kinds of options you can use → how to set up those options.

**Note**: Themes (which are a type of plugin) are able to receive options from a site's `gatsby-config` to be used in its `gatsby-config` in order to allow themes to be composed together. This is done by exporting the `gatsby-config` as a function instead of an object. You can see an example of this in the [`gatsby-theme-blog`](https://github.com/gatsbyjs/themes/tree/master/packages/gatsby-theme-blog) and [`gatsby-theme-blog-core`](https://github.com/gatsbyjs/themes/tree/master/packages/gatsby-theme-blog-core) repositories. Plugins are not capable of this functionality.

> ## How to validate options

Author: This entire section was moved to its own page, and (on the website) is liked right at the end of this page since it is the next page defined in `doc-links.yml`
